corehq/apps/locations/tests/test_location_sync.py

```diff
@@ -0,0 +1,118 @@
+    loc = SQLLocation(site_code=name, name=name, domain=DOMAIN, parent=parent,
+                      location_type=location_type, **OTHER_FIELDS)
+    loc.save()
+    return loc
+
+
+class TestLocationSync(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.domain_obj = create_domain(DOMAIN)
+        loc_types = setup_location_types(DOMAIN, ['state', 'county', 'city'])
+        cls.state = loc_types['state']
+        cls.county = loc_types['county']
+        cls.city = loc_types['city']
+        cls.db = Location.get_db()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.domain_obj.delete()
+
+    def tearDown(self):
+        delete_all_locations()
+
+    def assertLocationsEqual(self, loc1, loc2):
+        fields = ["domain", "name", "location_id", "location_type_name",
+                  "site_code", "external_id", "metadata", "is_archived"]
+        for field in fields:
+            msg = "The locations have different values for '{}'".format(field)
+            self.assertEqual(getattr(loc1, field), getattr(loc2, field), msg)
+
+        # <type 'Decimal'> != <type 'float'>
+        self.assertEqual(float(loc1.latitude), float(loc2.latitude))
+        self.assertEqual(float(loc1.longitude), float(loc2.longitude))
+
+        def get_parent(loc):
+            return loc.parent.location_id if loc.parent else None
+        self.assertEqual(get_parent(loc1), get_parent(loc2))
+
+    def assertNumLocations(self, number):
+        self.assertEqual(SQLLocation.objects.count(), number)
+        self.assertEqual(get_doc_count_by_type(self.db, 'Location'), number)
+
+    def test_sync_couch_to_sql(self):
+        mass = couch_loc("Massachusetts", self.state)
+        suffolk = couch_loc("Suffolk", self.state, mass)
+        boston = couch_loc("Boston", self.state, suffolk)
+        self.assertNumLocations(3)
+
+        for loc in [boston, suffolk, mass]:
+            self.assertLocationsEqual(loc, loc.sql_location)
+            loc.delete()
+            with self.assertRaises(SQLLocation.DoesNotExist):
+                SQLLocation.objects.get(location_id=loc.location_id)
+        self.assertNumLocations(0)
+
+    def test_save_couch_without_loc_type(self):
+        with self.assertRaises(SQLLocation.location_type.RelatedObjectDoesNotExist):
+            Location(site_code="no-type", name="no-type", domain=DOMAIN).save()
+        self.assertNumLocations(0)
+
+    def test_sync_sql_to_couch(self):
+        mass = sql_loc("Massachusetts", self.state)
+        suffolk = sql_loc("Suffolk", self.state, mass)
+        boston = sql_loc("Boston", self.state, suffolk)
+        self.assertNumLocations(3)
+
+        for loc in [boston, suffolk, mass]:
+            self.assertLocationsEqual(loc, loc.couch_location)
+            loc.delete()
+            with self.assertRaises(ResourceNotFound):
+                Location.get(loc.location_id)
+        self.assertNumLocations(0)
+
+    def test_edit_sql(self):
```

Comment:
i typically like to check for counts in the edit tests too, to make sure it's not creating any more records than it should